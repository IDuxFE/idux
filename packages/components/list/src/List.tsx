/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, normalizeClass, provide } from 'vue'

import { getFirstValidNode } from '@idux/cdk/utils'
import { ɵEmpty } from '@idux/components/_private/empty'
import { useGlobalConfig } from '@idux/components/config'
import { IxRow } from '@idux/components/grid'
import { IxSpin } from '@idux/components/spin'

import { listToken } from './token'
import { ListProps, listProps } from './types'

export default defineComponent({
  name: 'IxList',
  props: listProps,
  setup(props: ListProps, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-list`)
    const listConfig = useGlobalConfig('list')

    provide(listToken, {
      props,
      mergedPrefixCls,
    })

    const loadingProp = computed(() => {
      if (typeof props.loading === 'boolean') {
        return {
          spinning: props.loading,
        }
      }
      return props.loading
    })

    const classes = computed(() => {
      const prefixCls = mergedPrefixCls.value
      const borderless = props.borderless ?? listConfig.borderless
      const size = props.size ?? listConfig.size
      return normalizeClass({
        [`${prefixCls}`]: true,
        [`${prefixCls}-split`]: true,
        [`${prefixCls}-${size}`]: true,
        [`${prefixCls}-border`]: !borderless,
      })
    })

    return () => {
      const header = props.header ?? slots.header?.()
      const footer = props.footer ?? slots.footer?.()
      const children = slots.default ? slots.default() : []
      const content = getFirstValidNode(children) ? children : <ɵEmpty v-slots={slots} empty={props.empty} />

      const renderContent = () => {
        if (props.grid) {
          return <IxRow gutter={props.grid.gutter}>{content}</IxRow>
        }
        return <div>{content}</div>
      }

      return (
        <div class={classes.value}>
          {header && <div class={`${mergedPrefixCls.value}-header`}>{header}</div>}
          <IxSpin {...loadingProp.value}>{renderContent()}</IxSpin>
          {footer && <div class={`${mergedPrefixCls.value}-footer`}>{footer}</div>}
        </div>
      )
    }
  },
})
