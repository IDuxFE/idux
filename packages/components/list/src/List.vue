<template>
  <div class="ix-list" :class="classes">
    <div v-if="isShowHeader" class="ix-list-header">
      <slot name="header"> {{ header }} </slot>
    </div>
    <ix-spin :spinning="loading">
      <list-wrap :isUseGrid="isUseGrid" :gutter="grid?.gutter">
        <slot></slot>
      </list-wrap>
      <div v-show="isShowEmpty" class="ix-list-empty">
        <slot name="empty">
          <ix-empty :description="empty"></ix-empty>
        </slot>
      </div>
      <div v-if="isShowLoadMore" class="ix-list-loadMore">
        <slot name="loadMore">
          <ix-button :loading="loadMoreLoading" @click="handleLoadMoreClick">{{ loadMore }}</ix-button>
        </slot>
      </div>
    </ix-spin>
    <div v-if="isShowFooter" class="ix-list-footer">
      <slot name="footer"> {{ footer }} </slot>
    </div>
  </div>
</template>
<script lang="ts">
import type { ListProps } from './types'

import { computed, defineComponent, ref, provide } from 'vue'
import { IxSpin } from '@idux/components/spin'
import { IxButton } from '@idux/components/button'
import { IxEmpty } from '@idux/components/empty'
import { ListConfig, useGlobalConfig } from '@idux/components/config'
import ListWrap from './ListWrap.vue'
import { listProps } from './types'
import { listToken } from './token'

export default defineComponent({
  name: 'IxList',
  components: {
    ListWrap,
    IxSpin,
    IxButton,
    IxEmpty,
  },
  props: listProps,
  emits: ['loadMore'],
  setup(props: ListProps, { slots, emit }) {
    provide(
      listToken,
      computed(() => props.grid),
    )

    const isUseGrid = computed(() => !!props.grid)
    const isShowEmpty = computed(() => !slots.default)
    const isShowLoadMore = computed(() => slots.loadMore || props.loadMore)
    const isShowHeader = computed(() => slots.header || props.header)
    const isShowFooter = computed(() => slots.footer || props.footer)
    const listConfig = useGlobalConfig('list')
    const classes = useClasses(props, listConfig)

    const loadMoreLoading = ref(false)
    const handleLoadMoreClick = () => {
      loadMoreLoading.value = true
      const done = () => {
        loadMoreLoading.value = false
      }
      emit('loadMore', done)
    }

    return {
      isUseGrid,
      isShowEmpty,
      isShowLoadMore,
      isShowHeader,
      isShowFooter,
      loadMoreLoading,
      handleLoadMoreClick,
      classes,
    }
  },
})

const useClasses = (props: ListProps, listConfig: ListConfig) => {
  return computed(() => {
    const borderless = props.borderless ?? listConfig.borderless
    const size = props.size ?? listConfig.size ?? 'medium'
    const split = props.split
    return [
      `ix-list-${size}`,
      split ? 'ix-list-split' : '',
      {
        'ix-list-border': !borderless,
      },
    ]
  })
}
</script>
