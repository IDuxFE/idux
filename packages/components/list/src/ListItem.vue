<template>
  <list-item-wrap :grid="gird">
    <div class="ix-list-item-title">
      <slot name="title">{{ title }}</slot>
    </div>
    <div class="ix-list-item-content">
      <slot>{{ content }}</slot>
    </div>
    <div class="ix-list-item-extra">
      <slot name="extra">{{ extra }}</slot>
    </div>
  </list-item-wrap>
</template>
<script lang="ts">
import type { ListGridProps } from './types'

import { defineComponent, inject } from 'vue'
import ListItemWrap from './ListItemWrap.vue'
import { listToken } from './token'
import { listItemProps } from './types'

export default defineComponent({
  name: 'IxListItem',
  components: { ListItemWrap },
  props: listItemProps,
  setup() {
    const listGrid = inject(listToken, null)
    const gird = listGrid?.value && useGrid(listGrid?.value)
    return { gird }
  },
})

const getGrid = <T, K extends keyof T>(grid: T, dim: K) => {
  const v = grid[dim]
  if (typeof v === 'number') {
    return Math.floor(24 / v)
  }

  return undefined
}

const useGrid = (grid: ListGridProps) => {
  const defaultSpan = getGrid(grid, 'column')
  const xsSpan = getGrid(grid, 'xs')
  const smSpan = getGrid(grid, 'sm')
  const mdSpan = getGrid(grid, 'md')
  const lgSpan = getGrid(grid, 'lg')
  const xlSpan = getGrid(grid, 'xl')
  return {
    xs: xsSpan || defaultSpan,
    sm: smSpan || defaultSpan,
    md: mdSpan || defaultSpan,
    ld: lgSpan || defaultSpan,
    xl: xlSpan || defaultSpan,
  }
}
</script>
