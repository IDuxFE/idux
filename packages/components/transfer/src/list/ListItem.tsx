/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { VNodeChild, computed, defineComponent, inject, normalizeClass } from 'vue'

import { callEmit } from '@idux/cdk/utils'
import { IxCheckbox } from '@idux/components/checkbox'
import { IxIcon } from '@idux/components/icon'
import { convertStringVNode } from '@idux/components/utils'

import { transferListContext } from '../token'
import { transferListItemProps } from '../types'

export default defineComponent({
  props: transferListItemProps,
  setup(props, { slots }) {
    const { mergedPrefixCls } = inject(transferListContext)!

    const onCheckChange = (value: number | boolean | string) => {
      callEmit(props.onCheckChange, !!value)
    }
    const onRemove = () => {
      callEmit(props.onRemove)
    }

    const classes = computed(() => {
      const prefixCls = `${mergedPrefixCls.value}-item`

      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-disabled`]: props.disabled,
      })
    })

    return () => {
      const prefixCls = `${mergedPrefixCls.value}-item`

      const children: VNodeChild[] = []
      if (props.checkable) {
        children.push(
          <IxCheckbox
            class={`${prefixCls}-checkbox`}
            v-slots={slots}
            label={props.label}
            value={props.value}
            checked={props.checked}
            disabled={props.disabled}
            onChange={onCheckChange}
          />,
        )
      } else {
        const labelNode = convertStringVNode(slots.default, props.label)
        children.push(<label class={`${prefixCls}-label`}>{labelNode}</label>)
      }

      if (props.removable && !props.disabled) {
        children.push(
          <div class={`${prefixCls}-close-icon`} onClick={onRemove}>
            <IxIcon name="close" />
          </div>,
        )
      }

      return <li class={classes.value}>{children}</li>
    }
  },
})
