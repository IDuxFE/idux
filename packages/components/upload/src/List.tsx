/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { UploadListProps } from './types'
import type { UploadListConfig } from '@idux/components/config'

import { computed, defineComponent, h, inject, watchEffect } from 'vue'

import { useGlobalConfig } from '@idux/components/config'

import IxUploadImageCardList from './component/ImageCardList'
import IxUploadImageList from './component/ImageList'
import IxUploadTextList from './component/TextList'
import { useSelectorVisible } from './composables/useDisplay'
import { UploadToken, uploadToken } from './token'
import { uploadListProps } from './types'

const cpmMap = {
  text: IxUploadTextList,
  image: IxUploadImageList,
  imageCard: IxUploadImageCardList,
} as const

export default defineComponent({
  name: 'IxUploadList',
  props: uploadListProps,
  setup(props) {
    const config = useGlobalConfig('uploadList')
    const listType = useListType(props, config)
    const { props: uploadProps, setSelectorVisible } = inject(uploadToken, {
      props: {},
      setSelectorVisible: () => {},
    } as unknown as UploadToken)
    const [outerSelector] = useSelectorVisible(uploadProps, listType)

    watchEffect(() => setSelectorVisible(outerSelector.value))

    return () => h(cpmMap[listType.value], { ...props })
  },
})

function useListType(props: UploadListProps, config: UploadListConfig) {
  return computed(() => props.type ?? config.listType)
}
