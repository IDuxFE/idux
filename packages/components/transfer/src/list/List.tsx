/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, normalizeClass, normalizeStyle, provide, ref } from 'vue'

import { CdkVirtualScroll, type VirtualItemRenderFn, type VirtualScrollInstance } from '@idux/cdk/scroll'
import { callEmit, convertCssPixel } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'

import { transferListContext } from '../token'
import { type TransferData, type TransferListApi, transferListProps } from '../types'
import ListItem from './ListItem'

export default defineComponent({
  name: 'IxTransferList',
  props: transferListProps,
  setup(props, { slots, expose }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-transfer-list`)
    const labelKey = computed(() => props.labelKey ?? 'label')
    const virtualScrollRef = ref<VirtualScrollInstance>()

    provide(transferListContext, {
      mergedPrefixCls,
    })

    const transferListApi: TransferListApi = {
      scrollTo: option => virtualScrollRef.value?.scrollTo(option),
    }

    expose(transferListApi)

    const getKey = (item: TransferData) => (props.getKey?.(item) ?? item.key)!

    const handleScroll = (evt: Event) => {
      callEmit(props.onScroll, evt)
    }
    const handleScrolledBottom = () => {
      callEmit(props.onScrolledBottom)
    }
    const handleScrolledChange = (startIndex: number, endIndex: number, visibleData: unknown[]) => {
      callEmit(props.onScrolledChange, startIndex, endIndex, visibleData)
    }

    const renderListItem: VirtualItemRenderFn<TransferData> = ({ item, index }) => {
      const key = getKey(item)
      const onCheckChange = (checked: boolean) => {
        callEmit(props.onCheckChange, item, checked)
      }
      const onRemove = () => {
        callEmit(props.onRemove, item)
      }

      const customAdditional = props.customAdditional ? props.customAdditional({ data: item, index }) : undefined

      return (
        <ListItem
          key={key}
          value={key!}
          label={item[labelKey.value] as string}
          checked={!!props.checked?.(item)}
          disabled={!!props.disabled?.(item)}
          checkable={props.checkable}
          removable={props.removable}
          v-slots={{ default: slots.label && (() => slots.label?.(item)) }}
          onCheckChange={onCheckChange}
          onRemove={onRemove}
          {...(item.additional ?? {})}
          {...customAdditional}
        />
      )
    }

    const renderBody = () => {
      const { dataSource, virtual, scroll } = props
      const data = dataSource ?? []

      if (data.length <= 0) {
        return
      }

      if (virtual) {
        return (
          <CdkVirtualScroll
            ref={virtualScrollRef}
            dataSource={data}
            fullHeight={!!scroll?.fullHeight}
            getKey={getKey}
            height={(scroll?.height as number) ?? '100%'}
            itemHeight={32}
            itemRender={renderListItem}
            virtual
            onScroll={handleScroll}
            onScrolledBottom={handleScrolledBottom}
            onScrolledChange={handleScrolledChange}
          />
        )
      }

      return (
        <ul class={`${mergedPrefixCls.value}-inner`}>{data.map((item, index) => renderListItem({ item, index }))}</ul>
      )
    }

    const classes = computed(() => {
      const prefixCls = mergedPrefixCls.value

      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-virtual`]: !!props.virtual,
      })
    })
    const style = computed(() => {
      const scroll = props.scroll

      return (
        scroll &&
        normalizeStyle({
          [scroll.fullHeight ? 'height' : 'maxHeight']: convertCssPixel(scroll.height),
        })
      )
    })

    return () => {
      return (
        <div class={classes.value} style={style.value} onScroll={handleScroll}>
          {renderBody()}
        </div>
      )
    }
  },
})
