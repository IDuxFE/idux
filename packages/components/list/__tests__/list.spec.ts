import { flushPromises, mount } from '@vue/test-utils'
import List from '../src/List.vue'
import ListItem from '../src/ListItem.vue'
import { renderWork } from '@tests'

const TestComponent = {
  components: { List, ListItem },
  template: `
  <List 
    :size="size" 
    :header="header" 
    :footer="footer" 
    :loadMore="loadMore" 
    :split="split"
    :loading="loading"
    :grid="grid"
    :borderless="borderless"
    @loadMore="handleLoadMore">
    <ListItem>default</ListItem>
    <ListItem>default</ListItem>
    <ListItem>default</ListItem>
    <ListItem>default</ListItem>
  </List>
  `,
  props: {
    size: String,
    header: String,
    footer: String,
    loadMore: String,
    split: Boolean,
    loading: Boolean,
    grid: Object,
    borderless: Boolean,
    handleLoadMore: Function,
  },
}

async function sleep(time: number) {
  return new Promise(resolve => {
    setTimeout(resolve, time)
  })
}

describe('List.vue', () => {
  renderWork(TestComponent)

  test('size', async () => {
    const wrapper = mount(TestComponent)

    expect(wrapper.findAll('.ix-list-medium').length).toBe(1)
    expect(wrapper.findAll('.ix-list-small').length).toBe(0)
    expect(wrapper.findAll('.ix-list-large').length).toBe(0)

    await wrapper.setProps({ size: 'large' })

    expect(wrapper.findAll('.ix-list-medium').length).toBe(0)
    expect(wrapper.findAll('.ix-list-small').length).toBe(0)
    expect(wrapper.findAll('.ix-list-large').length).toBe(1)
  })

  test('header', async () => {
    const wrapper = mount(TestComponent)

    expect(wrapper.find('.ix-list-header').exists()).toBeFalsy()
    await wrapper.setProps({ header: 'header' })
    expect(wrapper.find('.ix-list-header').text()).toEqual('header')
    expect(wrapper.find('.ix-list-header').exists()).toBeTruthy()
  })
  test('footer', async () => {
    const wrapper = mount(TestComponent)

    expect(wrapper.find('.ix-list-footer').exists()).toBeFalsy()
    await wrapper.setProps({ footer: 'footer' })
    expect(wrapper.find('.ix-list-footer').text()).toEqual('footer')
    expect(wrapper.find('.ix-list-footer').exists()).toBeTruthy()
  })
  test('loadMore & spin', async () => {
    const wrapper = mount(TestComponent)

    expect(wrapper.find('.ix-list-loadMore').exists()).toBeFalsy()
    await wrapper.setProps({
      loadMore: '123',
      handleLoadMore: async function (done: jest.DoneCallback) {
        await sleep(1000)
        done()
      },
    })
    expect(wrapper.find('.ix-list-loadMore').exists()).toBeTruthy()
    const loadMoreEL = wrapper.element.querySelector('.ix-list-loadMore') as HTMLElement
    const loadMoreBtn = loadMoreEL.querySelector('button') as HTMLElement
    loadMoreBtn.click()
    await flushPromises()
    expect(wrapper.find('.ix-button-loading').exists()).toBeTruthy()
    await sleep(1100)
    expect(wrapper.find('.ix-button-loading').exists()).toBeFalsy()
    await wrapper.setProps({ loading: true })
    expect(wrapper.find('.ix-spin-spinner').exists()).toBeTruthy()
  })
  test('split', async () => {
    const wrapper = mount(TestComponent)

    expect(wrapper.classes()).not.toContain('ix-list-split')
    await wrapper.setProps({ split: true })
    expect(wrapper.classes()).toContain('ix-list-split')
  })
  test('borderless', async () => {
    const wrapper = mount(TestComponent)

    expect(wrapper.classes()).toContain('ix-list-border')
    await wrapper.setProps({ borderless: true })
    expect(wrapper.classes()).not.toContain('ix-list-borderless')
    await wrapper.setProps({ borderless: '' })
    expect(wrapper.classes()).not.toContain('ix-list-borderless')
  })

  test('grid', async () => {
    const wrapper = mount(TestComponent)

    expect(wrapper.find('.ix-row').exists()).toBeFalsy()
    await wrapper.setProps({
      grid: {
        column: 2,
        gutter: [10, 10],
        xs: 2,
      },
    })
    expect(wrapper.find('.ix-row').exists()).toBeTruthy()
    expect(wrapper.find('.ix-col').exists()).toBeTruthy()
    expect(wrapper.find('.ix-list-item').exists()).toBeFalsy()
  })
})
