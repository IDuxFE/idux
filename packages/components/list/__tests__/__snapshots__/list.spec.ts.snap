// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List render work 1`] = `
"<div class=\\"ix-list ix-list-medium ix-list-border\\">
  <!--v-if-->
  <div class=\\"ix-spin\\">
    <!---->
    <div class=\\"ix-spin-container\\">
      <div>
        <div class=\\"ix-list-item\\">
          <div class=\\"ix-list-item-title\\"></div>
          <div class=\\"ix-list-item-content\\">default</div>
          <div class=\\"ix-list-item-extra\\"></div>
        </div>
        <div class=\\"ix-list-item\\">
          <div class=\\"ix-list-item-title\\"></div>
          <div class=\\"ix-list-item-content\\">default</div>
          <div class=\\"ix-list-item-extra\\"></div>
        </div>
        <div class=\\"ix-list-item\\">
          <div class=\\"ix-list-item-title\\"></div>
          <div class=\\"ix-list-item-content\\">default</div>
          <div class=\\"ix-list-item-extra\\"></div>
        </div>
        <div class=\\"ix-list-item\\">
          <div class=\\"ix-list-item-title\\"></div>
          <div class=\\"ix-list-item-content\\">default</div>
          <div class=\\"ix-list-item-extra\\"></div>
        </div>
      </div>
      <div class=\\"ix-list-empty\\" style=\\"display: none;\\">
        <div class=\\"ix-empty\\">
          <div class=\\"ix-empty-image\\"><i class=\\"ix-icon ix-icon-empty\\" role=\\"img\\" aria-label=\\"empty\\"></i></div>
          <div class=\\"ix-empty-description\\">暂无数据</div>
          <!--v-if-->
        </div>
      </div>
      <!--v-if-->
    </div>
  </div>
  <!--v-if-->
</div>"
`;
